tinygrad/tensor.py

```diff
@@ -617,6 +617,30 @@ def dropout(self, p=0.5) -> Tensor:
     mask = (Tensor.rand(*self.shape, requires_grad=False) >= p).cast(dtypes.bool)
     return self * mask * (1/(1.0 - p))
 
+  # note: the way onehot is implemented, if value is out of range, all 0s are returned
+  def onehot(self, num_classes) -> Tensor: 
+    assert dtypes.is_int(self.dtype), f"onehot only supported for ints, got {self.dtype}"
+    return self.reshape(list(self.shape)+[1]).repeat([1]*len(self.shape)+[num_classes]).eq(Tensor.arange(num_classes, dtype=dtypes.int32)) # might not be ideal type, could be bool or int64 (convenience when type is passed further)
+  
+  def negative_log_likelihood(self, target:Tensor, weight:Optional[Tensor] = None, ignore_index=-100, reduction:Literal['none', 'mean', 'sum'] = 'mean') -> Tensor: 
+    assert target.shape == self.shape[:-1], f"Y dimensions {target.shape} must match all except last self dimension {self.shape}"
+    assert weight is None or (len(weight.shape) == 1 and weight.shape[0] == self.shape[-1]), f"weight dimensions {weight.shape} must match last self dimension {self.shape[-1]}"
+    # label_valid = (Y >= 0).mul(Y < self.shape[-1]).add(Y == ignore_index).minimum(1).min().cast(dtypes.bool)
+    # assert label_valid == True, "Cross entropy label out of range or not 'ignore_index' value" # cant check without realizing tensor :(  
+    num_classes = self.shape[-1]
+    y = target.onehot(num_classes)
+    if weight is not None:
+      W = Tensor.eye(num_classes, dtype=weight.dtype) * weight
```

Comment:
probably not a big deal in this example, but what would generally be a good choice of dtype in these cases?